File: api/reprompt/services/turn1_fitness.py
"""Turn 1 fitness scorer: optimized as conversation opener for max citation rate."""

import re


def score(query: str) -> int:
    s = 0
    q = query.lower().strip()

    # +40: Factual grounding patterns that demand web search
    grounding = [
        r"\bwhat are the (best|top|most)\b",
        r"\bwhich \w+ (is|are)\b",
        r"\btop[- ]rated\b",
        r"\bbest \w+ for\b",
        r"\bwhat is the best\b",
    ]
    if any(re.search(p, q) for p in grounding):
        s += 40

    # +30: Self-contained (no prior-context references)
    context_refs = [
        r"\bthat one\b",
        r"\bthe above\b",
        r"\bthose\b",
        r"\bthe one you\b",
        r"\bas you (said|mentioned)\b",
        r"^it\b",
    ]
    if not any(re.search(p, q) for p in context_refs):
        s += 30

    # +30: Research-journey initiator (broad + specific enough)
    words = q.split()
    if 4 < len(words) < 20 and re.search(r"\b(for|best|top|rated|recommended)\b", q):
        s += 30

    # -40: Follow-up/clarification patterns
    followup = [
        r"\bcan you explain\b",
        r"\btell me more\b",
        r"\bwhat about\b",
        r"\bexpand on\b",
        r"\bgo (deeper|further)\b",
        r"\belaborate\b",
    ]
    if any(re.search(p, q) for p in followup):
        s -= 40

    return max(0, min(100, s))

File: web/src/lib/scoring/turn1-fitness.ts
// Turn 1 fitness scorer: optimized as conversation opener for max citation rate.

export function score(query: string): number {
  let s = 0;
  const q = query.toLowerCase().trim();

  const grounding = [
    /\bwhat are the (best|top|most)\b/,
    /\bwhich \w+ (is|are)\b/,
    /\btop[- ]rated\b/,
    /\bbest \w+ for\b/,
    /\bwhat is the best\b/,
  ];
  if (grounding.some(p => p.test(q))) s += 40;

  const contextRefs = [/\bthat one\b/, /\bthe above\b/, /\bthose\b/, /\bthe one you\b/, /\bas you (said|mentioned)\b/, /^it\b/];
  if (!contextRefs.some(p => p.test(q))) s += 30;

  const words = q.split(/\s+/);
  if (words.length > 4 && words.length < 20 && /\b(for|best|top|rated|recommended)\b/.test(q)) s += 30;

  const followup = [/\bcan you explain\b/, /\btell me more\b/, /\bwhat about\b/, /\bexpand on\b/, /\bgo (deeper|further)\b/, /\belaborate\b/];
  if (followup.some(p => p.test(q))) s -= 40;

  return Math.max(0, Math.min(100, s));
}
